README.md
Bika Health LIS
===============

Bika Laboratory Information Management System customised for use in Health laboratories.

Last stable version: **v3.2** · released: 2015-11-03

[**Naringenin**](http://en.wikipedia.org/wiki/Naringenin) is a flavanone, a typeof flavonoid, that is considered to have a bioactive effect on human health as antioxidant, free radical scavenger, anti-inflammatory, carbohydrate metabolism promoter, and immune system modulator. It is the predominant flavanone in grapefruit.

Installation and upgrades
-------------------------
Bika Health can be installed on Windows, Mac OS X, Linux, BSD and other platforms:
https://github.com/bikalabs/bika.health/wiki/Installing-Bika-Health-Extension

Follow the instructions [Installing Bika Health Extension](https://github.com/bikalabs/bika.health/wiki/Installing-Bika-Health-Extension).

You might also find useful the following recipes:
- Installing Bika LIMS: https://github.com/bikalabs/Bika-LIMS/wiki/Bika-LIMS-Installation
- Installing Plone: https://plone.org/documentation/manual/installing-plone
- Installing Plone add-ons: https://plone.org/documentation/kb/installing-add-ons-quick-how-to

Documentation
-------------
- Bika Health Wiki: http://github.com/bikalabs/bika.health/wiki
- Bika LIMS User Manual: http://demo.bikalabs.com/knowledge-centre/manual/bika-3-user-manual
- Bika LIMS Wiki: http://github.com/bikalabs/Bika-LIMS/wiki

Feedback and support
--------------------
If you have questions, please post to one of our mailing lists:

* Bika Health Users: http://groups.google.com/forum/?hl=en#!forum/bika-health
* Bika LIMS Users: http://lists.sourceforge.net/lists/listinfo/bika-users
* LIMS design: https://groups.google.com/forum/?hl=en#%21forum/bika-design
* Developers: http://lists.sourceforge.net/lists/listinfo/bika-developers

Please log issues, feature requests, or bug reports in the [issue tracker](http://jira.bikalabs.com/)

Professional services and support
---------------------------------
* Naralabs, http://naralabs.com
* Bika Lab Systems, http://bikalabs.com


docs/CHANGELOG.txt
3.1.9 (unreleased)
------------------
HEALTH-357: UI. Post Patient Create landing page
HEALTH-503: Validation of Date of birth 
HEALTH-356: SMS Results Reporting and Notifications
HEALTH-413: Dashboard not showing
HEALTH-402: In Analysis Request Add form, contact doesn't get selected

3.1.8 (2015-11-03)
------------------
HEALTH-301: Use newer version of Health flow diagram in next release
HEALTH-281: Error in Bika Setup › Insurance Companies > Invoices
HEALTH-305: Having invoices
HEALTH-370: Worksheet architecture has changed in LIMS
HEALTH-269: Analysis Request Add compatibility with LIMS 3.1.9
HEALTH-273: Error upgrading to 317
HEALTH-270: Error while importing patient with "yearinprefix" disabled
HEALTH-271: Analysis request invoice view broken
HEALTH-266: View error on invoice from analysis request
HEALTH-258: Add "File attachment" on Patient

3.1.7 (2015-06-09)
------------------
HEALTH-282: Error loading Add Analysis request
HEALTH-245: Set-up data load. Patient ID conversion, alternatives
HEALTH-227: Converting Patient IDs before import
HEALTH-228: Load Setup data bugs
HEALTH-140: AR Create per path lab standard form
HEALTH-251: Add guarantor details in insurance companies

3.1.6 (2015-02-27)
------------------
HEALTH-223: When you are adding a doctor through an overlay (add doctor button in cases), the address widgets don't work properly.
HEALTH-215: Correct Navigation tree order
HEALTH-191: Client Contact permissions
HEALTH-137: Medical Insurance for Patients. Alternative invoice workflow
HEALTH-204: Doctor Samples view broken
HEALTH-200: Additional picklists don't work when creating Patients directly from the Case's view
HEALTH-136: [+ Add] Patient button on AR Create form.
HEALTH-179: "Copy to new" button in AR, doesn't copy the selected data.
HEALTH-197: Health's results report error.
HEALTH-204: Batch.samples error loading page
HEALTH-177: Past Medical History and Drug History's end date selection error.
HEALTH-178: Past Medical History, Travell History and Immunization History data: Impossible to remove the last set of data.
HEALTH-208: Incompatibilty with new Bika LIMS' add site templates
HEALTH-197: Health's results report error
HEALTH-189: Patient Edit page: After define a country, it is not saved.
HEALTH-184: Add Case: Patient Age at Case Onset Date doesn't get filled after introduce the Onset Date.


3.1.5 (2014-10-10)
------------------

HEALTH-176: Cannot choose a drug from the dropdown list on Patient->allergies
HEALTH-174: Unable to add an AR to a Case (missing Client)
HEALTH-163: Cannot be chosen Immunisation items from dropdown list.
HEALTH-162: Cannot choose Drug Prohibition Explanation
HEALTH-169: Analyst can't see Samples site, Admin can.
HEALTH-168: Diagnosis ICD widget does not complete from Code, Description
HEALTH-161: In Add Patient, after introducing the Birth Date, the patient's age don't get filled automatically.
HEALTH-157: Patient field is missing in AR add views
HEALTH-150: Compatibility with the new JS loader machinery
HEALTH-164: Editing a patient, location not have to be important on Travel History field
HEALTH-166: "Patient Age at Case Onset Date" in Add Case cannot be filled
HEALTH-172: Anything is displayed on drugs list
HEALTH-173: ImmunizationHistory cannot save data
HEALTH-149: Compatibility with the new Bika LIMS reporting subsystem
HEALTH-152: Upgrade the test data worksheet
HEALTH-145: Health icons not used. And 1 more
HEALTH-152: Upgrade the test data worksheet

- Plus Bika-LIMS 3.1.5: http://git.io/ogjDuQ


3.1.2.1 (2014-08-05)
------------------

HEALTH-144: Set up data: Identifier Types NameError: global name '_id' is not defined
HEALTH-143: Can not create site


3.1.2 (2014-07-25)
------------------

HEALTH-104: Health Setup data failures
HEALTH-28: Health load setup data. Drugs and Treatments did not import
HEALTH-105: Case syndromic classifications site eror in setup
HEALTH-93: AR. Updating Info portion
HEALTH-27: AR Create. Copying Patient fields across does not autocomplete corresponding Name or ID

- Plus Bika-LIMS 3.1.2 and 3.1.3: http://git.io/MWb4dQ


3.1.1 (2014-07-11)
------------------

HEALTH-122: Client contact cannot open Client page/AR page blank
HEALTH-92: Display Patients tab inside Referral Institution (Client)
HEALTH-109: Manually adding symptoms saving but remaining on the same page
HEALTH-124: Client contact can access doctors and patients of other clients
HEALTH-121: Client contact gets Insufficient Privileges upon login
HEALTH-133: Case creation: Basal body temperature fields accepting values way out of range

bika/health/content/patient.py
from Products.ATContentTypes.content import schemata
from AccessControl import ClassSecurityInfo
from Products.ATContentTypes.utils import DT2dt
from Products.ATExtensions.ateapi import RecordsField
from Products.Archetypes import atapi
from Products.Archetypes.public import *
from Products.CMFCore.utils import getToolByName
from Products.CMFPlone.utils import safe_unicode
from bika.lims import bikaMessageFactory as _b
from bika.health import bikaMessageFactory as _
from bika.lims.browser.fields import AddressField
from bika.lims.browser.widgets import AddressWidget
from bika.lims.browser.widgets import DateTimeWidget
from bika.lims.browser.widgets import RecordsWidget
from bika.health.widgets import SplittedDateWidget
from bika.health.config import *
from bika.lims.content.person import Person
from bika.health.interfaces import IPatient
from bika.health.permissions import *
from bika.health.widgets import ReadonlyStringWidget
from datetime import datetime
from zope.interface import implements
from bika.health.widgets.patientmenstrualstatuswidget import PatientMenstrualStatusWidget

schema = Person.schema.copy() + Schema((
    StringField('PatientID',
                searchable=1,
                required=0,
                widget=ReadonlyStringWidget(
                    visible={'view': 'visible', 'edit': 'hidden'},
                    label=_('Patient ID'),
                    css='readonly-emphasize',
                ),
    ),
    ReferenceField('PrimaryReferrer',
                   vocabulary='get_clients',
                   allowed_types=('Client',),
                   relationship='PatientClient',
                   required=1,
                   widget=SelectionWidget(
                       format='select',
                       label=_('Client'),
                   ),
    ),
    ComputedField('PrimaryReferrerTitle',
                  expression="context.Schema()['PrimaryReferrer'].get(context) and context.Schema()['PrimaryReferrer'].get(context).Title() or None",
                  widget=ComputedWidget(
                  ),
    ),
     ComputedField('PrimaryReferrerUID',
                  expression="context.Schema()['PrimaryReferrer'].get(context) and context.Schema()['PrimaryReferrer'].get(context).UID() or None",
                  widget=ComputedWidget(
                  ),
    ),
    StringField('Gender',
                vocabulary=GENDERS,
                index='FieldIndex',
                default='dk',
                widget=SelectionWidget(
                    format='select',
                    label=_('Gender'),
                ),
    ),
    StringField('Age',
                widget=StringWidget(
                    label=_('Age'),
                    visible=0,
                    width=3,
                ),
    ),
    DateTimeField('BirthDate',
                  required=0,
                  widget=DateTimeWidget(
                      label=_('Birth date'),
                  ),
    ),
    BooleanField('BirthDateEstimated',
                 default=False,
                 widget=BooleanWidget(
                     label=_('Birth date is estimated'),
                 ),
    ),
    RecordsField('AgeSplitted',
                 required=1,
                 widget=SplittedDateWidget(
                     label=_('Age'),
                 ),
    ),
    AddressField('CountryState',
                 widget=AddressWidget(
                 label=_("Country and state"),
                     showLegend=True,
                     showDistrict=False,
                     showCopyFrom=False,
                     showCity=False,
                     showPostalCode=False,
                     showAddress=False,
                 ),
    ),
    TextField('Remarks',
              searchable=True,
              default_content_type='text/plain',
              allowable_content_types = ('text/plain', ),
              default_output_type="text/plain",
                  widget=TextAreaWidget(
                      macro="bika_widgets/remarks",
                  label=_('Remarks'),
                  append_only=True,
              ),
    ),
    RecordsField('TreatmentHistory',
                 type='treatmenthistory',
                 subfields=('Treatment',
                            'Drug',
                            'Start',
                            'End'),
                 required_subfields=('Drug',
                                     'Start',
                                     'End'),
                 subfield_labels={'Drug': _('Drug'),
                                  'Start': _('Start'),
                                  'End': _('End')},
                 subfield_sizes={'Drug': 40,
                                 'Start': 10,
                                 'End': 10},
                 subfield_types={'Start': 'datepicker_nofuture',
                                 'End': 'datepicker'},
                 widget=RecordsWidget(
                 label='Drug History',
                 description=_("A list of patient treatments and drugs administered."),
                     combogrid_options={
                         'Drug': {
                             'colModel': [{'columnName':'Title', 'width':'30', 'label':_('Title')},
                                          {'columnName':'Description', 'width':'70', 'label':_('Description')}],
                             'url': 'getdrugs',
                             'showOn': True,
                             'width': '550px'
                         },
                     },
                 ),
    ),
    RecordsField('Allergies',
                 type='allergies',
                 subfields=('DrugProhibition',
                            'Drug',
                            'Remarks'),
                 required_subfields=('DrugProhibition',
                                     'Drug'),
                 subfield_labels={'DrugProhibition': _('Drug Prohibition Explanation'),
                                  'Drug': _('Drug'),
                                  'Remarks': _('Remarks')},
                 subfield_sizes={'DrugProhibition': 30,
                                 'Drug': 30,
                                 'Remarks': 30},
                 widget=RecordsWidget(
                     label='Allergies',
                     description=_("Known Patient allergies to keep information that can aid drug reaction interpretation"),
                     combogrid_options={
                         'Drug': {
                             'colModel': [{'columnName':'Title', 'width':'30', 'label':_('Title')},
                                          {'columnName':'Description', 'width':'70', 'label':_('Description')}],
                             'url': 'getdrugs',
                             'showOn': True,
                             'width': '550px'
                         },
                         'DrugProhibition': {
                             'colModel': [{'columnName':'DrugProhibition', 'width':'30', 'label':_('Title')},
                                          {'columnName':'Description', 'width':'70', 'label':_('Description')}],
                             'url': 'getdrugprohibitions',
                             'showOn': True,
                             'width': '550px'
                         },
                     },
                 ),
    ),
    RecordsField('ImmunizationHistory',
                 type='immunizationhistory',
                 subfields=('EPINumber',
                            'Immunization',
                            'VaccinationCenter',
                            'Date',
                            'Remarks'),
                 required_subfields=('EPINumber',
                                     'Immunization',
                                     'Date'),
                 subfield_labels={'EPINumber': _('EPI Number'),
                                  'Immunization': _('Immunization'),
                                  'VaccinationCenter': _('Vaccination Center'),
                                  'Date': _('Date'),
                                  'Remarks': _('Remarks')},
                 subfield_sizes={'EPINumber': 12,
                                 'Immunization': 20,
                                 'VaccinationCenter': 10, 'Date': 10, 'Remarks': 25},
                 subfield_types={'Date':'datepicker_nofuture'},
                 widget=RecordsWidget(
                     label='Immunization History',
                     description=_("A list of immunizations administered to the patient."),
                     combogrid_options={
                         'Immunization': {
                             'colModel': [{'columnName':'Immunization', 'width':'30', 'label':_('Title')},
                                         {'columnName':'Description', 'width':'70', 'label':_('Description')}],
                             'url': 'getimmunizations',
                             'showOn': True,
                             'width': '550px'
                         },
                         'VaccinationCenter': {
                             'colModel': [{'columnName':'VaccinationCenter', 'width':'100', 'label':_('Title')}],
                             'url': 'getvaccinationcenters',
                             'showOn': True,
                             'width': '550px'
                         },
                     },
                 ),
    ),
    RecordsField('TravelHistory',
                 type='travelhistory',
                 subfields=('TripStartDate',
                            'TripEndDate',
                            'Country',
                            'Location',
                            'Remarks'),
                 required_subfields=('Country'),
                 subfield_labels={'TripStartDate': _('Trip Start Date', 'Start date'),
                                  'TripEndDate': _('Trip End Date', 'End date'),
                                  'Country': _('Country'),
                                  'Location': _('Location'),
                                  'Remarks': _('Remarks')},
                 subfield_sizes={'TripStartDate': 10,
                                 'TripEndDate': 10,
                                 'Country': 20,
                                 'Location': 20,
                                 'Remarks': 25},
                 subfield_types={'TripStartDate':'datepicker_nofuture',
                                 'TripEndDate':'datepicker'},
                 widget=RecordsWidget(
                     label='Travel History',
                     description=_("A list of places visited by the patient."),
                     combogrid_options={
                         'Country': {
                             'colModel': [{'columnName':'Code', 'width':'15', 'label':_('Code')},
                                          {'columnName':'Country', 'width':'85', 'label':_('Country')}],
                             'url': 'getCountries',
                             'showOn': True,
                             'width': "450px",
                         },
                     },
                 ),
    ),
    RecordsField('ChronicConditions',
                 type='chronicconditions',
                 subfields=('Code',
                            'Title',
                            'Description',
                            'Onset',
                            'End'),
                 required_subfields=('Title',
                                     'Onset'),
                 subfield_sizes={'Code': 7,
                                 'Title': 20,
                                 'Description': 35,
                                 'Onset': 10,
                                 'End': 10},
                 subfield_types={'Onset': 'datepicker_nofuture',
                                 'End': 'datepicker'},
                 widget=RecordsWidget(
                     label='Past Medical History',
                     description=_("Patient's past medical history."),
                     combogrid_options={
                         'Title': {
                             'colModel': [{'columnName':'Code', 'width':'10', 'label':_('Code')},
                                          {'columnName':'Title', 'width':'30', 'label':_('Title')},
                                          {'columnName':'Description', 'width':'60', 'label':_('Description')}],
                             'url': 'getdiseases',
                             'showOn': True,
                             'width': "650px",
                         },
                     },
                 ),
    ),
    StringField('BirthPlace', schemata='Personal',
                widget=StringWidget(
                    label=_('Birth place'),
                ),
    ),
    StringField('Ethnicity', schemata='Personal',
                index='FieldIndex',
                vocabulary=ETHNICITIES,
                widget=ReferenceWidget(
                    label=_('Ethnicity'),
                    description=_("Ethnicity eg. Asian, African, etc."),
                ),
    ),
    StringField('Citizenship', schemata='Personal',
                widget=StringWidget(
                    label=_('Citizenship'),
                ),
    ),
    StringField('MothersName', schemata='Personal',
                widget=StringWidget(
                    label=_('Mothers name'),
                ),
    ),
    StringField('CivilStatus', schemata='Personal',
                widget=StringWidget(
                    label=_('Civil status'),
                ),
    ),
    ImageField('Photo', schemata='Identification',
               widget=ImageWidget(
                   label=_('Photo'),
               ),
    ),
    ImageField('Feature', schemata='Identification',
               multiValue=1,
               widget=ImageWidget(
                   label=_('Feature'),
               ),
    ),
    RecordsField('MenstrualStatus',
            type='menstrualstatus',
            widget=PatientMenstrualStatusWidget(
                label='Menstrual status',
            ),
    ),
    StringField('ClientPatientID',
            searchable=1,
            required=0,
            widget=StringWidget(
                label=_('Client Patient ID'),
            ),
    ),
    BooleanField('Anonymous',
             default=False,
             widget=BooleanWidget(
                 label=_("Anonymous")
             ),
    ),
    BooleanField('DefaultResultsDistribution',
        schemata="Publication preference",
        default=True,
        widget=BooleanWidget(
            label=_("Inherit default settings"),
            description=_("If checked, settings will be inherited from "
                          "the Client, so further changes in Client for this "
                          "setting will be populated too."))
    ),
    BooleanField('AllowResultsDistribution',
        schemata="Publication preference",
        default=False,
        widget=BooleanWidget(
            label=_("Allow results distribution to this patient"),
            description=_("If checked, results reports will also be sent "
                          "to the Patient automatically."))
    ),
    LinesField('PublicationPreferences',
        vocabulary=PUBLICATION_PREFS,
        schemata='Publication preference',
        widget=MultiSelectionWidget(
            label=_("Publication preference"),
            description=_("Select the preferred channels to be used for "
                          "sending the results reports to this Patient."))
    ),
    BooleanField('PublicationAttachmentsPermitted',
        default=False,
        schemata='Publication preference',
        widget=BooleanWidget(
            label=_("Results attachments permitted"),
            description=_("File attachments to results, e.g. microscope "
                          "photos, will be included in emails to patient "
                          "if this option is enabled"))
    ),
))

schema['JobTitle'].widget.visible = False
schema['Department'].widget.visible = False
schema['BusinessPhone'].widget.visible = False
schema['BusinessFax'].widget.visible = False
# Don't make title required - it will be computed from the Person's Fullname
schema['title'].required = 0
schema['title'].widget.visible = False
schema['EmailAddress'].schemata = 'Personal'
schema['HomePhone'].schemata = 'Personal'
schema['MobilePhone'].schemata = 'Personal'
#schema.moveField('PatientID', pos='top')
schema.moveField('PrimaryReferrer', after='Surname')
schema.moveField('PatientID', before='title')
schema.moveField('Identifier', after='PrimaryReferrer')
schema.moveField('Gender', after='Identifier')
schema.moveField('Age', after='Gender')
schema.moveField('BirthDate', after='Age')
schema.moveField('BirthDateEstimated', after='BirthDate')
schema.moveField('AgeSplitted', after='BirthDateEstimated')
schema.moveField('CountryState', after='AgeSplitted')
schema.moveField('MenstrualStatus', after='AgeSplitted')
schema.moveField('ClientPatientID', after='PatientID')
schema.moveField('Anonymous', before='ClientPatientID')


class Patient(Person):
    implements(IPatient)
    security = ClassSecurityInfo()
    displayContentsTab = False
    schema = schema

    _at_rename_after_creation = True

    def _renameAfterCreation(self, check_auto_id=False):
        from bika.lims.idserver import renameAfterCreation
        renameAfterCreation(self)

    def Title(self):
        """ Return the Fullname as title """
        return safe_unicode(self.getFullname()).encode('utf-8')

    security.declarePublic('getPatientID')

    def getPatientID(self):
        return self.getId()

    security.declarePublic('getSamples')
    def getSamples(self):
        """ get all samples taken from this Patient """
        return [ar.getSample() for ar in self.getARs() if ar.getSample()]

    security.declarePublic('getARs')
    def getARs(self, analysis_state=None):
        bc = getToolByName(self, 'bika_catalog')
        ars = bc(portal_type='AnalysisRequest')
        outars = []
        for ar in ars:
            ar = ar.getObject()
            pat = ar.Schema()['Patient'].get(ar) if ar else None
            if pat and pat.UID() == self.UID():
                outars.append(ar)
        return outars

    def get_clients(self):
        ## Only show clients to which we have Manage AR rights.
        mtool = getToolByName(self, 'portal_membership')
        clientfolder = self.clients
        clients = []
        for client in clientfolder.objectValues("Client"):
            if not mtool.checkPermission(ManageAnalysisRequests, client):
                continue
            clients.append([client.UID(), client.Title()])
        clients.sort(lambda x, y: cmp(x[1].lower(), y[1].lower()))
        clients.insert(0, ['', ''])
        return DisplayList(clients)

    def getIdentifierStr(self):
        ids = self.getIdentifier()
        idsstr = ''
        for id in ids:
            idsstr += idsstr == '' and id['Identifier'] or (', ' + id['Identifier'])
        return idsstr
        #return self.getSendersPatientID()+" "+self.getSendersCaseID()+" "+self.getSendersSpecimenID()

    def getIdentifierStrHtml(self):
        ids = self.getIdentifier()
        idsstr = '<table cellpadding="0" cellspacing="0" border="0" class="patientsidentifiers" style="text-align:left;width: 100%;"><tr><td>'
        for id in ids:
            idsstr += "<tr><td>" + id['IdentifierType'] + ':</td><td>' + id['Identifier'] + "</td></tr>"
        return "</table>" + idsstr

    def getAgeSplitted(self):

        if (self.getBirthDate()):
            dob = DT2dt(self.getBirthDate()).replace(tzinfo=None)
            now = datetime.today()

            currentday = now.day
            currentmonth = now.month
            currentyear = now.year
            birthday = dob.day
            birthmonth = dob.month
            birthyear = dob.year
            ageday = currentday - birthday
            agemonth = 0
            ageyear = 0
            months31days = [1, 3, 5, 7, 8, 10, 12]

            if (ageday < 0):
                currentmonth -= 1
                if (currentmonth < 1):
                    currentyear -= 1
                    currentmonth = currentmonth + 12

                dayspermonth = 30
                if currentmonth in months31days:
                    dayspermonth = 31
                elif currentmonth == 2:
                    dayspermonth = 28
                    if(currentyear % 4 == 0
                       and (currentyear % 100 > 0 or currentyear % 400 == 0)):
                        dayspermonth += 1

                ageday = ageday + dayspermonth

            agemonth = currentmonth - birthmonth
            if (agemonth < 0):
                currentyear -= 1
                agemonth = agemonth + 12

            ageyear = currentyear - birthyear

            return [{'year': ageyear,
                    'month': agemonth,
                    'day': ageday}]
        else:
            return [{'year': '',
                    'month': '',
                    'day': ''}]

    def getAge(self):
        return self.getAgeSplitted()[0]['year']

    def getAgeSplittedStr(self):
        splitted = self.getAgeSplitted()[0]
        arr = []
        arr.append(splitted['year'] and str(splitted['year']) + 'y' or '')
        arr.append(splitted['month'] and str(splitted['month']) + 'm' or '')
        arr.append(splitted['day'] and str(splitted['day']) + 'd' or '')
        return ' '.join(arr)

    def setCountryState(self, value):
        pa = self.getPhysicalAddress() and self.getPhysicalAddress() or {'country': '', 'state': ''}
        pa['country'] = self.REQUEST.form.get('CountryState', {'country': ''})['country']
        pa['state'] = self.REQUEST.form.get('CountryState', {'state': ''})['state']
        if not pa['country']:
            return

        return self.setPhysicalAddress(pa)

    def getCountryState(self):
        return self.getPhysicalAddress()

# schemata.finalizeATCTSchema(schema, folderish=True, moveDiscussion=False)
atapi.registerType(Patient, PROJECTNAME)
